extensions/mssql/src/objectExplorerNodeProvider/webhdfs.ts

```diff
@@ -0,0 +1,717 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+'use strict';
+
+import * as util from 'util';
+import * as url from 'url';
+import * as fs from 'fs';
+import * as querystring from 'querystring';
+import * as request from 'request';
+import * as BufferStreamReader from 'buffer-stream-reader';
+import * as nls from 'vscode-nls';
+import { IHdfsOptions, IRequestParams } from './fileSources';
+
+const localize = nls.loadMessageBundle();
+
+export class WebHDFS {
+	private _requestParams: IRequestParams;
+	private _opts: IHdfsOptions;
+	private _url: any;
+
+	private static ErrorMessageInvalidDataStructure =
+		localize('webhdfs.invalidDataStructure', 'Invalid Data Structure');
+
+	constructor(opts: IHdfsOptions, requestParams: IRequestParams) {
+		if (!(this instanceof WebHDFS)) {
+			return new WebHDFS(opts, requestParams);
+		}
+
+		let missingProps = ['host', 'port', 'path']
+			.filter(p => !opts.hasOwnProperty(p));
+		if (missingProps && missingProps.length > 0) {
+			throw new Error(
+				util.format(
+					localize('webhdfs.missingProperties',
+						'Unable to create WebHDFS client: missing options: %s'),
+					missingProps.join(', ')
+				)
+			);
+		}
+
+		this._requestParams = requestParams;
+		this._opts = opts;
+		this._url = {
+			protocol: opts.protocol || 'http',
+			hostname: opts.host,
+			port: opts.port || 80,
+			pathname: opts.path
+		};
+	}
+
+	private checkArgUndefined(argName: string, argValue: any): void {
+		if (!argValue) {
+			throw new Error(
+				util.format(
+					localize('webhdfs.undefinedArgument', '\'%s\' is undefined.'),
+					argName
+				)
+			);
+		}
+	}
+
+	/**
+	 * Generate WebHDFS REST API endpoint URL for given operation
+	 *
+	 * @param {string} operation WebHDFS operation name
+	 * @param {string} path
+	 * @param {object} params
+	 * @returns {string} WebHDFS REST API endpoint URL
+	 */
+	private getOperationEndpoint(operation: string, path: string, params?: object): string {
+		let endpoint = this._url;
+		endpoint.pathname = this._opts.path + path;
+		let searchOpts = Object.assign(
+			{ 'op': operation },
+			this._opts.user ? { 'user.name': this._opts.user } : {},
+			params || {}
+		);
+		endpoint.search = querystring.stringify(searchOpts);
+		return url.format(endpoint);
+	}
+
+	/**
+	 * Gets localized status message for given status code
+	 *
+	 * @param {number} statusCode Http status code
+	 * @returns {string} Status message
+	 */
+	private toStatusMessage(statusCode: number): string {
```

Comment:
What if the error message has additional information? Shouldn't we append after the generate error information?

Reply:
It seems like we should get as much information as possible available.

Reply:
fixed